agent/consul/state/connect_ca.go

```diff
@@ -157,7 +159,7 @@ func (s *Store) CACheckAndSetConfig(idx, cidx uint64, config *structs.CAConfigur
 	// return early here.
 	e, ok := existing.(*structs.CAConfiguration)
 	if (ok && e.ModifyIndex != cidx) || (!ok && cidx != 0) {
-		return false, nil
+		return false, errors.Errorf("invalid index")
```

Comment:
Hmm I prefer returning an explicit error although I think the wording could be a little different - this is somewhat 'expected' in that the user requested "Check and Set" so it's not that their index is _invalid_ so much as that their requested check has failed and their write may have conflicted with another one.

Perhaps a message like `CAS failure: ModifyIndex did not match existing` would be clearer and also easier for consumers to detect since their behviour in this case might be to retry the read-modify-write rather than this being a fatal error.

Reply:
Are there any other callers to this other than the `CAManager` one we modified above?

Reply:
This is only triggered by leader_connect_ca.go in 3 places:
- UpdateConfiguration (the one we are fixing)
- initializeCAConfig which read the config from raft copy it and modify it and write it again
- initializeRootCA which add the root ca to the config and write it again

Reply:
Don't forget that `Store.CACheckAndSetConfig` is also called by the `CAOpSetRootsAndConfig` operation, which is called in two other places:
* a second place in `UpdateConfiguration`
* `CAManager.persistNewRootAndConfig` called by `initializeSecondaryCA` and `RenewIntermediate`, although `RenewIntermediate` will use a `nil` config, so should always use the existing stored config and not hit this bug.